There is a failing test right now due to the lack of #1544:

```
======================================================================
ERROR: Ensure that the package commiters can be found using the Pagure
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/home/vagrant/bodhi/bodhi/tests/server/test_models.py", line 417, in test_get_pkg_committers_from_pagure
    rv = self.package.get_pkg_committers_from_pagure()
  File "/home/vagrant/bodhi/bodhi/server/models.py", line 584, in get_pkg_committers_from_pagure
    config['pagure_url'].rstrip('/'), group_name)
  File "/home/vagrant/bodhi/bodhi/server/config.py", line 47, in __getitem__
    return super(BodhiConfig, self).__getitem__(*args, **kw)
KeyError: 'pagure_url'
```

It's up to you whether you want to make this code work the old way so we can merge it before #1544, or whether you want to wait for me to merge #1544 so we can rebase this PR on it and update it to use the new central config system. I'd recommend the latter, but I'm OK with either you prefer.

--- a/.gitignore
+++ b/.gitignore
@@ -26,3 +26,4 @@ html-docs
 masher/
 Vagrantfile
 .dnf-cache/
+.idea

--- a/bodhi/server/models.py
+++ b/bodhi/server/models.py
@@ -43,7 +43,7 @@ from bodhi.server import bugs, buildsys, log, mail, notifications, Session
 from bodhi.server.config import config
 from bodhi.server.exceptions import BodhiException, LockedUpdateException
 from bodhi.server.util import (
-    avatar as get_avatar, build_evr, flash_log, get_critpath_pkgs,
+    avatar as get_avatar, build_evr, flash_log, get_critpath_components,
     get_nvr, get_rpm_header, header, tokenize, pagure_api_get)
 import bodhi.server.util
 
@@ -639,7 +639,7 @@ class Package(Base):
             # the former.
             for group_name in package_json['access_groups'][access_type]:
                 group_pagure_url = '{0}/api/0/group/{1}'.format(
-                    config['pagure_url'].rstrip('/'), group_name)
+                    pagure_url.rstrip('/'), group_name)
                 group_json = pagure_api_get(group_pagure_url)
                 committers = committers | set(group_json['members'])
 
@@ -1201,6 +1201,26 @@ class Update(Base):
                 break
             yield comment
 
+    @staticmethod
+    def contains_critpath_component(builds, release_name):
+        """
+        Determines if there is a critpath component in the builds passed in.
+        :param builds: a list of Build objects
+        :param release_name: a string representing the name of the release
+        such as "f25"
+        :return: a boolean determining if the update contains a critpath
+        component
+        """
+        for build in builds:
+            # This function call is cached, so there is no need to optimize
+            # it here.
+            critpath_components = get_critpath_components(
+                release_name.lower(), build.package.type.value)
+            if build.package.name in critpath_components:
+                return True
+
+        return False
+
     @classmethod
     def new(cls, request, data):
         """ Create a new update """
@@ -1208,18 +1228,9 @@ class Update(Base):
         user = User.get(request.user.name, request.db)
         data['user'] = user
         data['title'] = ' '.join([b.nvr for b in data['builds']])
-
         caveats = []
-
-        critical = False
-        critpath_pkgs = get_critpath_pkgs(data['release'].name.lower())
-        if critpath_pkgs:
-            for build in data['builds']:
-                if build.package.name in critpath_pkgs:
-                    critical = True
-                    break
-
-        data['critpath'] = critical
+        data['critpath'] = cls.contains_critpath_component(
+            data['builds'], data['release'].name)
 
         # Create the Bug entities, but don't talk to rhbz yet.  We do that
         # offline in the UpdatesHandler fedmsg consumer now.
@@ -1326,15 +1337,8 @@ class Update(Base):
                     # an override
                     db.delete(b)
 
-        critical = False
-        critpath_pkgs = get_critpath_pkgs(up.release.name.lower())
-        if critpath_pkgs:
-            for build in up.builds:
-                if build.package.name in critpath_pkgs:
-                    critical = True
-                    break
-
-        data['critpath'] = critical
+        data['critpath'] = cls.contains_critpath_component(
+            up.builds, up.release.name)
 
         del(data['builds'])
 

--- a/bodhi/server/models.py
+++ b/bodhi/server/models.py
@@ -44,7 +44,7 @@ from bodhi.server.config import config
 from bodhi.server.exceptions import BodhiException, LockedUpdateException
 from bodhi.server.util import (
     avatar as get_avatar, build_evr, flash_log, get_critpath_pkgs,
-    get_nvr, get_rpm_header, header, tokenize)
+    get_nvr, get_rpm_header, header, tokenize, pagure_api_get)
 import bodhi.server.util
 
 
@@ -611,6 +611,45 @@ class Package(Base):
                             committers.append(name)
 
         return (committers, watchers), (committergroups, watchergroups)
+
+    def get_pkg_committers_from_pagure(self):
+        """
+        Pull users who can commit and are watching a package in Pagure.
+
+        Return two two-tuples of lists:
+        * The first tuple is for usernames.  The second tuple is for groups.
+        * The first list of the tuple is for committers. The second is for
+          watchers.
+        """
+        pagure_url = config.get(
+            'pagure_url', 'https://src.fedoraproject.org/pagure/')
+        # Pagure uses plural names for its namespaces such as "rpms"
+        namespace = self.type.name + 's'
+        package_pagure_url = '{0}/api/0/{1}/{2}'.format(
+            pagure_url.rstrip('/'), namespace, self.name)
+        package_json = pagure_api_get(package_pagure_url)
+
+        committers = set()
+        for access_type in ['owner', 'admin', 'commit']:
+            committers = committers | set(
+                package_json['access_users'][access_type])
+        for access_type in ['admin', 'commit']:
+            # We resolve the group membership here since these are Pagure
+            # groups, and not FAS groups. Bodhi understands the latter but not
+            # the former.
+            for group_name in package_json['access_groups'][access_type]:
+                group_pagure_url = '{0}/api/0/group/{1}'.format(
+                    config['pagure_url'].rstrip('/'), group_name)
+                group_json = pagure_api_get(group_pagure_url)
+                committers = committers | set(group_json['members'])
+
+        # The first list contains users with commit access. The second list is
+        # supposed to contain groups with commit access. Since Pagure uses
+        # Pagure groups, which Bodhi doesn't understand, the code above
+        # resolves the group membership. That is why the second list is empty.
+        # These return values maintain compatibility with other functions such
+        # as get_pkg_pushers.
+        return list(committers), []
 
     def fetch_test_cases(self, db):
         """ Get a list of test cases from the wiki """

--- a/bodhi/server/util.py
+++ b/bodhi/server/util.py
@@ -167,21 +167,32 @@ class memoized(object):
 
 
 @memoized
-def get_critpath_pkgs(collection='master'):
-    """Return a list of critical path packages for a given collection"""
-    critpath_pkgs = []
+def get_critpath_components(collection='master', component_type='rpm'):
+    """ Return a list of critical path packages for a given collection.
+    :param collection: a string representing the collection/branch to search.
+    for
+    :param component_type: a string representing the component type to search
+    for. This only affects PDC queries.
+    :return: a list of critpath components
+    """
+    critpath_components = []
     critpath_type = config.get('critpath.type')
+    if critpath_type != 'pdc' and component_type != 'rpm':
+        log.warning('The critpath.type of "{0}" does not support searching for'
+                    ' non-RPM components'.format(component_type))
+
     if critpath_type == 'pkgdb':
         from pkgdb2client import PkgDB
         pkgdb = PkgDB(config.get('pkgdb_url'))
         results = pkgdb.get_critpath_packages(branches=collection)
         if collection in results['pkgs']:
-            critpath_pkgs = results['pkgs'][collection]
+            critpath_components = results['pkgs'][collection]
     elif critpath_type == 'pdc':
-        critpath_pkgs = get_critpath_pkgs_from_pdc(collection)
+        critpath_components = get_critpath_components_from_pdc(
+            collection, component_type)
     else:
-        critpath_pkgs = config.get('critpath_pkgs', '').split()
-    return critpath_pkgs
+        critpath_components = config.get('critpath_pkgs', '').split()
+    return critpath_components
 
 
 class Singleton(object):
@@ -659,7 +670,7 @@ def sort_severity(value):
     return value_map.get(value, 99)
 
 
-def get_critpath_pkgs_from_pdc(branch):
+def get_critpath_components_from_pdc(branch, component_type='rpm'):
     """
     Searches PDC for critical path packages based on the specified branch.
     :param branch: string representing the branch
@@ -674,7 +685,7 @@ def get_critpath_pkgs_from_pdc(branch):
         'critical_path': 'true',
         'name': branch,
         'page_size': 100,
-        'type': 'rpm'
+        'type': component_type
     })
     pdc_api_url_with_args = '{0}?{1}'.format(pdc_api_url, query_args)
 

--- a/bodhi/server/util.py
+++ b/bodhi/server/util.py
@@ -177,6 +177,8 @@ def get_critpath_pkgs(collection='master'):
         results = pkgdb.get_critpath_packages(branches=collection)
         if collection in results['pkgs']:
             critpath_pkgs = results['pkgs'][collection]
+    elif critpath_type == 'pdc':
+        critpath_pkgs = get_critpath_pkgs_from_pdc(collection)
     else:
         critpath_pkgs = config.get('critpath_pkgs', '').split()
     return critpath_pkgs
@@ -655,3 +657,96 @@ def sort_severity(value):
     }
 
     return value_map.get(value, 99)
+
+
+def get_critpath_pkgs_from_pdc(branch):
+    """
+    Searches PDC for critical path packages based on the specified branch.
+    :param branch: string representing the branch
+    :param component_type: string representing the component type to get
+    critpath packages for. This only affects queries to PDC.
+    :return: a list of critical path packages
+    """
+    pdc_api_url = '{}/rest_api/v1/component-branches/'.format(
+        config.get('pdc_url', 'https://pdc.fedoraproject.org/').rstrip('/'))
+    query_args = urllib.urlencode({
+        'active': 'true',
+        'critical_path': 'true',
+        'name': branch,
+        'page_size': 100,
+        'type': 'rpm'
+    })
+    pdc_api_url_with_args = '{0}?{1}'.format(pdc_api_url, query_args)
+
+    critpath_pkgs_set = set()
+    while True:
+        pdc_request_json = pdc_api_get(pdc_api_url_with_args)
+
+        for branch_rv in pdc_request_json['results']:
+            critpath_pkgs_set.add(branch_rv['global_component'])
+
+        if pdc_request_json['next']:
+            pdc_api_url_with_args = pdc_request_json['next']
+        else:
+            # There are no more results to iterate through
+            break
+    return list(critpath_pkgs_set)
+
+
+def api_get(api_url, service_name, error_key=None):
+    """
+    A wrapper to get API resources with some error handling.
+    :param api_url: a string of the URL to query
+    :param service_name: a string of the service name to use in error messages
+    :param error_key: a string of the key that holds the error message in the
+    JSON body. If this is set to None, the whole JSON will be used as the
+    error message.
+    :return: dictionary of the returned JSON or a RuntimeError
+    """
+    base_error_msg = (
+        'Bodhi failed to get a resource from {0} at the following URL '
+        '"{1}". The status code was "{2}".')
+    rv = requests.get(api_url, timeout=60)
+
+    if rv.status_code == 200:
+        return rv.json()
+    elif rv.status_code == 500:
+        # There will be no JSON with an error message here
+        error_msg = base_error_msg.format(
+            service_name, api_url, rv.status_code)
+        log.error(error_msg)
+        raise RuntimeError(error_msg)
+    else:
+        # If it's not a 500 error, we can assume that the API returned an error
+        # message in JSON that we can log
+        try:
+            rv_error = rv.json()
+            if error_key is not None:
+                rv_error = rv_error.get(error_key)
+        except ValueError:
+            rv_error = ''
+        error_msg = base_error_msg.format(
+            service_name, api_url, rv.status_code)
+        error_msg = '{0} The error was "{1}".'.format(error_msg, rv_error)
+        log.error(error_msg)
+        raise RuntimeError(error_msg)
+
+
+def pagure_api_get(pagure_api_url):
+    """
+    A wrapper to get API resources from Pagure.
+    :param pagure_api_url: a string of the URL to query
+    :return: JSON of the API resource(s) or a RuntimeError
+    """
+    return api_get(pagure_api_url, service_name='Pagure', error_key='error')
+
+
+def pdc_api_get(pdc_api_url):
+    """
+    A wrapper to get API resources from PDC.
+    :param pdc_api_url: a string of the URL to query
+    :return: JSON of the API resource(s) or a RuntimeError
+    """
+    # There is no error_key specified because the error key is not consistent
+    # based on the error message
+    return api_get(pdc_api_url, service_name='PDC')

--- a/bodhi/server/validators.py
+++ b/bodhi/server/validators.py
@@ -369,6 +369,19 @@ def validate_acls(request):
         if has_access:
             continue
 
+        # Make sure the user is in the mandatory packager groups. This is a
+        # safeguard in the event a user has commit access on the ACL system
+        # but isn't part of the mandatory groups.
+        mandatory_groups = request.registry.settings[
+            'mandatory_packager_groups'].split()
+        for mandatory_group in mandatory_groups:
+            if mandatory_group not in user_groups:
+                error = ('{0} is not a member of "{1}", which is a '
+                         'mandatory packager group').format(
+                    user.name, mandatory_group)
+                request.errors.add('body', 'builds', error)
+                return
+
         if acl_system == 'pkgdb':
             try:
                 people, groups = package.get_pkg_pushers(
@@ -381,6 +394,24 @@ def validate_acls(request):
                                    "Unable to access the Package "
                                    "Database to check ACLs. Please "
                                    "try again later.")
+                return
+        elif acl_system == 'pagure':
+            try:
+                committers, groups = package.get_pkg_committers_from_pagure()
+                people = committers
+            except RuntimeError as error:
+                # If it's a RuntimeError, then the error will be logged
+                # and we can return the error to the user as is
+                log.error(error)
+                request.errors.add('body', 'builds', unicode(error))
+                return
+            except Exception as error:
+                # This is an unexpected error, so let's log it and give back
+                # a generic error to the user
+                log.exception(error)
+                error_msg = ('Unable to access Pagure to check ACLs. '
+                             'Please try again later.')
+                request.errors.add('body', 'builds', error_msg)
                 return
         elif acl_system == 'dummy':
             people = (['ralph', 'bowlofeggs', 'guest'], ['guest'])

--- a/bodhi/tests/server/services/test_updates.py
+++ b/bodhi/tests/server/services/test_updates.py
@@ -163,7 +163,14 @@ class TestNewUpdate(bodhi.tests.server.functional.base.BaseWSGICase):
         app = TestApp(main({}, testing=u'bodhi', session=self.db, **self.app_settings))
         res = app.post_json('/updates/', self.get_update(u'bodhi-2.1-1.fc17'),
                             status=400)
-        assert 'bodhi does not have commit access to bodhi' in res, res
+        expected_error = {
+            "location": "body",
+            "name": "builds",
+            "description": ("bodhi is not a member of \"packager\", which is a"
+                            " mandatory packager group")
+        }
+        assert expected_error in res.json_body['errors'], \
+            res.json_body['errors']
         self.assertEquals(publish.call_args_list, [])
 
     @mock.patch(**mock_taskotron_results)
@@ -503,11 +510,14 @@ class TestUpdatesService(bodhi.tests.server.functional.base.BaseWSGICase):
         nvr = u'bodhi-2.1-1.fc17'
 
         user = User(name=u'lloyd')
+        user2 = User(name=u'ralph')
         self.db.add(user)
-        self.db.add(User(name=u'ralph'))  # Add a non proventester
+        self.db.add(user2)  # Add a packager but not proventester
         self.db.flush()
         group = self.db.query(Group).filter_by(name=u'provenpackager').one()
         user.groups.append(group)
+        group2 = self.db.query(Group).filter_by(name=u'packager').one()
+        user2.groups.append(group2)
 
         app = TestApp(main({}, testing=u'ralph', session=self.db, **self.app_settings))
         up_data = self.get_update(nvr)
@@ -535,12 +545,15 @@ class TestUpdatesService(bodhi.tests.server.functional.base.BaseWSGICase):
         "Ensure provenpackagers can change the request for any update"
         nvr = u'bodhi-2.1-1.fc17'
         user = User(name=u'bob')
+        user2 = User(name=u'ralph')
         self.db.add(user)
-        self.db.add(User(name=u'ralph'))  # Add a non proventester
+        self.db.add(user2)  # Add a packager but not proventester
         self.db.add(User(name=u'someuser'))  # An unrelated user with no privs
         self.db.flush()
         group = self.db.query(Group).filter_by(name=u'provenpackager').one()
         user.groups.append(group)
+        group2 = self.db.query(Group).filter_by(name=u'packager').one()
+        user2.groups.append(group2)
 
         app = TestApp(main({}, testing=u'ralph', session=self.db, **self.app_settings))
         up_data = self.get_update(nvr)

--- a/bodhi/tests/server/test_models.py
+++ b/bodhi/tests/server/test_models.py
@@ -25,7 +25,7 @@ from sqlalchemy.exc import IntegrityError
 import cornice
 import mock
 
-from bodhi.server import models as model, buildsys, mail
+from bodhi.server import models as model, buildsys, mail, util
 from bodhi.server.config import config
 from bodhi.server.exceptions import BodhiException
 from bodhi.server.models import (
@@ -758,6 +758,97 @@ class TestUpdate(ModelTest):
             True)
         # No bugs should have been closed
         eq_(close.call_count, 0)
+
+    @mock.patch('bodhi.server.util.requests.get')
+    @mock.patch.dict(util.config, {
+        'critpath.type': 'pdc',
+        'pdc_url': 'http://domain.local'
+    })
+    def test_contains_critpath_component(self, mock_get):
+        """ Verifies that the static function of contains_critpath_component
+        determines that one of the builds has a critpath component.
+        """
+        mock_get.return_value.status_code = 200
+        mock_get.return_value.json.return_value = {
+            'count': 2,
+            'next': None,
+            'previous': None,
+            'results': [
+                {
+                    'active': True,
+                    'critical_path': True,
+                    'global_component': 'gcc',
+                    'id': 6,
+                    'name': 'f11',
+                    'slas': [],
+                    'type': 'rpm'
+                },
+                {
+                    'active': True,
+                    'critical_path': True,
+                    'global_component': 'TurboGears',
+                    'id': 7,
+                    'name': 'f11',
+                    'slas': [],
+                    'type': 'rpm'
+                }
+            ]
+        }
+        update = self.get_update()
+        self.assertTrue(update.contains_critpath_component(
+            update.builds, update.release.name))
+
+    @mock.patch('bodhi.server.util.requests.get')
+    @mock.patch.dict(util.config, {
+        'critpath.type': 'pdc',
+        'pdc_url': 'http://domain.local'
+    })
+    def test_contains_critpath_component_not_critpath(self, mock_get):
+        """ Verifies that the static function of contains_critpath_component
+        determines that none of the builds are critpath components.
+        """
+        mock_get.return_value.status_code = 200
+        mock_get.return_value.json.return_value = {
+            'count': 2,
+            'next': None,
+            'previous': None,
+            'results': [
+                {
+                    'active': True,
+                    'critical_path': True,
+                    'global_component': 'gcc',
+                    'id': 6,
+                    'name': 'f25',
+                    'slas': [],
+                    'type': 'rpm'
+                },
+                {
+                    'active': True,
+                    'critical_path': True,
+                    'global_component': 'python',
+                    'id': 7,
+                    'name': 'f25',
+                    'slas': [],
+                    'type': 'rpm'
+                }
+            ]
+        }
+        update = self.get_update()
+        # Use a different release here for additional testing and to avoid
+        # caching from the previous test
+        update.release = model.Release(
+            name=u'fc25', long_name=u'Fedora 25',
+            id_prefix=u'FEDORA', dist_tag=u'dist-fc25',
+            stable_tag=u'dist-fc25-updates',
+            testing_tag=u'dist-fc25-updates-testing',
+            candidate_tag=u'dist-fc25-updates-candidate',
+            pending_signing_tag=u'dist-fc25-updates-testing-signing',
+            pending_testing_tag=u'dist-fc25-updates-testing-pending',
+            pending_stable_tag=u'dist-fc25-updates-pending',
+            override_tag=u'dist-fc25-override',
+            branch=u'fc25', version=u'25')
+        self.assertFalse(update.contains_critpath_component(
+            update.builds, update.release.name))
 
     def test_unpush_build(self):
         eq_(len(self.obj.builds), 1)

--- a/bodhi/tests/server/test_models.py
+++ b/bodhi/tests/server/test_models.py
@@ -386,6 +386,62 @@ class TestRpmPackage(ModelTest, unittest.TestCase):
         # This should not raise any Exception.
         self.db.flush()
 
+    @mock.patch('bodhi.server.util.requests.get')
+    def test_get_pkg_committers_from_pagure(self, mock_get):
+        """ Ensure that the package commiters can be found using the Pagure
+        API.
+        """
+        mock_get.return_value.status_code = 200
+        json_one = {
+            "access_groups": {
+                "admin": [],
+                "commit": ['factory2'],
+                "ticket": []
+            },
+            "access_users": {
+                "admin": [],
+                "commit": [],
+                "owner": [
+                    "mprahl"
+                ],
+                "ticket": ["jsmith"]
+            },
+            "close_status": [],
+            "custom_keys": [],
+            "date_created": "1494947106",
+            "description": "Python",
+            "fullname": "rpms/python",
+            "id": 2,
+            "milestones": {},
+            "name": "python",
+            "namespace": "rpms",
+            "parent": None,
+            "priorities": {},
+            "tags": [],
+            "user": {
+                "fullname": "Matt Prahl",
+                "name": "mprahl"
+            }
+        }
+        json_two = {
+            "creator": {
+                "fullname": "Ralph Bean",
+                "name": "ralph"
+            },
+            "date_created": "1495035052",
+            "description": "Factory 2.0 Developers",
+            "display_name": "factory2",
+            "group_type": "user",
+            "members": [
+                "ralph",
+                "mikeb"
+            ],
+            "name": "factory2"
+        }
+        mock_get.return_value.json.side_effect = [json_one, json_two]
+        rv = self.package.get_pkg_committers_from_pagure()
+        assert rv == (['ralph', 'mikeb', 'mprahl'], []), rv
+
 
 class TestRpmBuild(ModelTest):
     """Unit test case for the ``RpmBuild`` model."""

--- a/bodhi/tests/server/test_utils.py
+++ b/bodhi/tests/server/test_utils.py
@@ -15,6 +15,7 @@ import subprocess
 import unittest
 
 import mock
+import pkgdb2client
 
 from bodhi.server import util
 from bodhi.server.buildsys import setup_buildsystem, teardown_buildsystem
@@ -33,17 +34,40 @@ class TestUtils(object):
         assert config.get('sqlalchemy.url'), config
         assert config['sqlalchemy.url'], config
 
-    def test_get_critpath_pkgs(self):
+    def test_get_critpath_components(self):
         """Ensure the pkgdb's critpath API works"""
-        pkgs = util.get_critpath_pkgs()
+        pkgs = util.get_critpath_components()
         assert 'kernel' in pkgs, pkgs
+
+    @mock.patch.object(pkgdb2client.PkgDB, 'get_critpath_packages')
+    @mock.patch.dict(util.config, {
+        'critpath.type': 'pkgdb',
+        'pkgdb_url': 'http://domain.local'
+    })
+    def test_get_critpath_components_pkgdb_success(self, mock_get_critpath):
+        """ Ensure that critpath packages can be found using PkgDB.
+        """
+        # A subset of critpath packages
+        critpath_pkgs = [
+            'pth',
+            'xorg-x11-server-utils',
+            'giflib',
+            'basesystem'
+        ]
+        mock_get_critpath.return_value = {
+            'pkgs': {
+                'f20': critpath_pkgs
+            }
+        }
+        pkgs = util.get_critpath_components('f20')
+        assert critpath_pkgs == pkgs, pkgs
 
     @mock.patch('bodhi.server.util.requests.get')
     @mock.patch.dict(util.config, {
         'critpath.type': 'pdc',
         'pdc_url': 'http://domain.local'
     })
-    def test_get_critpath_pkgs_pdc_error(self, mock_get):
+    def test_get_critpath_components_pdc_error(self, mock_get):
         """ Ensure an error is thrown in Bodhi if there is an error in PDC
         getting the critpath packages.
         """
@@ -51,7 +75,7 @@ class TestUtils(object):
         mock_get.return_value.json.return_value = \
             {'error': 'some error'}
         try:
-            util.get_critpath_pkgs('f25')
+            util.get_critpath_components('f25')
             assert False, 'Did not raise a RuntimeError'
         except RuntimeError as error:
             actual_error = unicode(error)
@@ -60,12 +84,26 @@ class TestUtils(object):
         assert 'Bodhi failed to get a resource from PDC' in actual_error
         assert 'The status code was "500".' in actual_error
 
+    @mock.patch('bodhi.server.util.log')
+    @mock.patch.dict(util.config, {
+        'critpath.type': 'dummy',
+    })
+    def test_get_critpath_components_not_pdc_not_rpm(self, mock_log):
+        """ Ensure a warning is logged when the critpath system is not pdc
+        and the type of components to search for is not rpm.
+        """
+        pkgs = util.get_critpath_components('f25', 'module')
+        assert 'kernel' in pkgs, pkgs
+        warning = ('The critpath.type of "module" does not support searching '
+                   'for non-RPM components')
+        mock_log.warning.assert_called_once_with(warning)
+
     @mock.patch('bodhi.server.util.requests.get')
     @mock.patch.dict(util.config, {
         'critpath.type': 'pdc',
         'pdc_url': 'http://domain.local'
     })
-    def test_get_critpath_pkgs_pdc_success(self, mock_get):
+    def test_get_critpath_components_pdc_success(self, mock_get):
         """ Ensure that critpath packages can be found using PDC.
         """
         pdc_url = \
@@ -106,7 +144,7 @@ class TestUtils(object):
                 ]
             }
         ]
-        pkgs = util.get_critpath_pkgs('f26')
+        pkgs = util.get_critpath_components('f26')
         assert 'python' in pkgs and 'gcc' in pkgs, pkgs
         # At least make sure it called the next url to cycle through the pages.
         # We can't verify all the calls made because the URL GET parameters

--- a/bodhi/tests/server/test_utils.py
+++ b/bodhi/tests/server/test_utils.py
@@ -38,6 +38,277 @@ class TestUtils(object):
         pkgs = util.get_critpath_pkgs()
         assert 'kernel' in pkgs, pkgs
 
+    @mock.patch('bodhi.server.util.requests.get')
+    @mock.patch.dict(util.config, {
+        'critpath.type': 'pdc',
+        'pdc_url': 'http://domain.local'
+    })
+    def test_get_critpath_pkgs_pdc_error(self, mock_get):
+        """ Ensure an error is thrown in Bodhi if there is an error in PDC
+        getting the critpath packages.
+        """
+        mock_get.return_value.status_code = 500
+        mock_get.return_value.json.return_value = \
+            {'error': 'some error'}
+        try:
+            util.get_critpath_pkgs('f25')
+            assert False, 'Did not raise a RuntimeError'
+        except RuntimeError as error:
+            actual_error = unicode(error)
+        # We are not testing the whole error message because there is no
+        # guarantee of the ordering of the GET parameters.
+        assert 'Bodhi failed to get a resource from PDC' in actual_error
+        assert 'The status code was "500".' in actual_error
+
+    @mock.patch('bodhi.server.util.requests.get')
+    @mock.patch.dict(util.config, {
+        'critpath.type': 'pdc',
+        'pdc_url': 'http://domain.local'
+    })
+    def test_get_critpath_pkgs_pdc_success(self, mock_get):
+        """ Ensure that critpath packages can be found using PDC.
+        """
+        pdc_url = \
+            'http://domain.local/rest_api/v1/component-branches/?page_size=1'
+        pdc_next_url = '{0}&page=2'.format(pdc_url)
+        mock_get.return_value.status_code = 200
+        mock_get.return_value.json.side_effect = [
+            {
+                'count': 2,
+                'next': pdc_next_url,
+                'previous': None,
+                'results': [
+                    {
+                        'active': True,
+                        'critical_path': True,
+                        'global_component': 'gcc',
+                        'id': 6,
+                        'name': 'f26',
+                        'slas': [],
+                        'type': 'rpm'
+                    }
+                ]
+            },
+            {
+                'count': 2,
+                'next': None,
+                'previous': pdc_url,
+                'results': [
+                    {
+                        'active': True,
+                        'critical_path': True,
+                        'global_component': 'python',
+                        'id': 7,
+                        'name': 'f26',
+                        'slas': [],
+                        'type': 'rpm'
+                    }
+                ]
+            }
+        ]
+        pkgs = util.get_critpath_pkgs('f26')
+        assert 'python' in pkgs and 'gcc' in pkgs, pkgs
+        # At least make sure it called the next url to cycle through the pages.
+        # We can't verify all the calls made because the URL GET parameters
+        # in the URL may have different orders based on the system/Python
+        # version.
+        mock_get.assert_called_with(pdc_next_url, timeout=60)
+        # Verify there were two GET requests made and two .json() calls
+        assert mock_get.call_count == 2, mock_get.call_count
+        assert mock_get.return_value.json.call_count == 2, \
+            mock_get.return_value.json.call_count
+
+    @mock.patch('bodhi.server.util.requests.get')
+    def test_pagure_api_get(self, mock_get):
+        """ Ensure that an API request to Pagure works as expected.
+        """
+        mock_get.return_value.status_code = 200
+        expected_json = {
+            "access_groups": {
+                "admin": [],
+                "commit": [],
+                "ticket": []
+            },
+            "access_users": {
+                "admin": [],
+                "commit": [],
+                "owner": [
+                    "mprahl"
+                ],
+                "ticket": []
+            },
+            "close_status": [],
+            "custom_keys": [],
+            "date_created": "1494947106",
+            "description": "Python",
+            "fullname": "rpms/python",
+            "id": 2,
+            "milestones": {},
+            "name": "python",
+            "namespace": "rpms",
+            "parent": None,
+            "priorities": {},
+            "tags": [],
+            "user": {
+                "fullname": "Matt Prahl",
+                "name": "mprahl"
+            }
+        }
+        mock_get.return_value.json.return_value = expected_json
+        rv = util.pagure_api_get('http://domain.local/api/0/rpms/python')
+        assert rv == expected_json, rv
+
+    @mock.patch('bodhi.server.util.requests.get')
+    def test_pagure_api_get_non_500_error(self, mock_get):
+        """ Ensure that an API request to Pagure that raises an error that is
+        not a 500 error returns the actual error message from the JSON.
+        """
+        mock_get.return_value.status_code = 404
+        mock_get.return_value.json.return_value = {
+            "error": "Project not found",
+            "error_code": "ENOPROJECT"
+        }
+        try:
+            util.pagure_api_get('http://domain.local/api/0/rpms/python')
+            assert False, 'Did not raise a RuntimeError'
+        except RuntimeError as error:
+            actual_error = unicode(error)
+
+        expected_error = (
+            'Bodhi failed to get a resource from Pagure at the following URL '
+            '"http://domain.local/api/0/rpms/python". The status code was '
+            '"404". The error was "Project not found".')
+        assert actual_error == expected_error, actual_error
+
+    @mock.patch('bodhi.server.util.requests.get')
+    def test_pagure_api_get_500_error(self, mock_get):
+        """ Ensure that an API request to Pagure that triggers a 500 error
+        raises the expected error message.
+        """
+        mock_get.return_value.status_code = 500
+        try:
+            util.pagure_api_get('http://domain.local/api/0/rpms/python')
+            assert False, 'Did not raise a RuntimeError'
+        except RuntimeError as error:
+            actual_error = unicode(error)
+
+        expected_error = (
+            'Bodhi failed to get a resource from Pagure at the following URL '
+            '"http://domain.local/api/0/rpms/python". The status code was '
+            '"500".')
+        assert actual_error == expected_error, actual_error
+
+    @mock.patch('bodhi.server.util.requests.get')
+    def test_pagure_api_get_non_500_error_no_json(self, mock_get):
+        """ Ensure that an API request to Pagure that raises an error that is
+        not a 500 error and has no JSON returns an error.
+        """
+        mock_get.return_value.status_code = 404
+        mock_get.return_value.json.side_effect = ValueError('Not JSON')
+        try:
+            util.pagure_api_get('http://domain.local/api/0/rpms/python')
+            assert False, 'Did not raise a RuntimeError'
+        except RuntimeError as error:
+            actual_error = unicode(error)
+
+        expected_error = (
+            'Bodhi failed to get a resource from Pagure at the following URL '
+            '"http://domain.local/api/0/rpms/python". The status code was '
+            '"404". The error was "".')
+        assert actual_error == expected_error, actual_error
+
+    @mock.patch('bodhi.server.util.requests.get')
+    def test_pdc_api_get(self, mock_get):
+        """ Ensure that an API request to PDC works as expected.
+        """
+        mock_get.return_value.status_code = 200
+        expected_json = {
+            "count": 1,
+            "next": None,
+            "previous": None,
+            "results": [
+                {
+                    "id": 1,
+                    "sla": "security_fixes",
+                    "branch": {
+                        "id": 1,
+                        "name": "2.7",
+                        "global_component": "python",
+                        "type": "rpm",
+                        "active": True,
+                        "critical_path": True
+                    },
+                    "eol": "2018-04-27"
+                }
+            ]
+        }
+        mock_get.return_value.json.return_value = expected_json
+        rv = util.pdc_api_get(
+            'http://domain.local/rest_api/v1/component-branch-slas/')
+        assert rv == expected_json, rv
+
+    @mock.patch('bodhi.server.util.requests.get')
+    def test_pdc_api_get_500_error(self, mock_get):
+        """ Ensure that an API request to PDC that triggers a 500 error
+        raises the expected error message.
+        """
+        mock_get.return_value.status_code = 500
+        try:
+            util.pdc_api_get(
+                'http://domain.local/rest_api/v1/component-branch-slas/')
+            assert False, 'Did not raise a RuntimeError'
+        except RuntimeError as error:
+            actual_error = unicode(error)
+
+        expected_error = (
+            'Bodhi failed to get a resource from PDC at the following URL '
+            '"http://domain.local/rest_api/v1/component-branch-slas/". The '
+            'status code was "500".')
+        assert actual_error == expected_error, actual_error
+
+    @mock.patch('bodhi.server.util.requests.get')
+    def test_pdc_api_get_non_500_error(self, mock_get):
+        """ Ensure that an API request to PDC that raises an error that is
+        not a 500 error returns the returned JSON.
+        """
+        mock_get.return_value.status_code = 404
+        mock_get.return_value.json.return_value = {
+            "detail": "Not found."
+        }
+        try:
+            util.pdc_api_get(
+                'http://domain.local/rest_api/v1/component-branch-slas/3/')
+            assert False, 'Did not raise a RuntimeError'
+        except RuntimeError as error:
+            actual_error = unicode(error)
+
+        expected_error = (
+            'Bodhi failed to get a resource from PDC at the following URL '
+            '"http://domain.local/rest_api/v1/component-branch-slas/3/". The '
+            'status code was "404". The error was '
+            '"{\'detail\': \'Not found.\'}".')
+        assert actual_error == expected_error, actual_error
+
+    @mock.patch('bodhi.server.util.requests.get')
+    def test_pdc_api_get_non_500_error_no_json(self, mock_get):
+        """ Ensure that an API request to PDC that raises an error that is
+        not a 500 error and has no JSON returns an error.
+        """
+        mock_get.return_value.status_code = 404
+        mock_get.return_value.json.side_effect = ValueError('Not JSON')
+        try:
+            util.pdc_api_get(
+                'http://domain.local/rest_api/v1/component-branch-slas/3/')
+            assert False, 'Did not raise a RuntimeError'
+        except RuntimeError as error:
+            actual_error = unicode(error)
+
+        expected_error = (
+            'Bodhi failed to get a resource from PDC at the following URL '
+            '"http://domain.local/rest_api/v1/component-branch-slas/3/". The '
+            'status code was "404". The error was "".')
+        assert actual_error == expected_error, actual_error
+
     def test_get_nvr(self):
         """Assert the correct return value and type from get_nvr()."""
         result = util.get_nvr(u'ejabberd-16.12-3.fc26')

--- a/bodhi/tests/server/test_validators.py
+++ b/bodhi/tests/server/test_validators.py
@@ -1,0 +1,175 @@
+# Copyright 2017 Red Hat, Inc.
+# This file is part of bodhi.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+"""This module contains tests for bodhi.server.validators."""
+import mock
+from cornice.errors import Errors
+
+from bodhi.server import validators
+from bodhi.tests.server.base import BaseTestCase
+from bodhi.server import models
+
+
+class TestValidateAcls(BaseTestCase):
+    """ Test the validate_acls() function.
+    """
+    def get_mock_request(self, acl_system):
+        """
+        A helper function that creates a mock request.
+        :param acl_system: a string representing the acl_system to set in the
+        configuration.
+        :return: a Mock object representing a request
+        """
+        update = self.db.query(models.Update).filter_by(
+            title='bodhi-2.0-1.fc17').one()
+        user = self.db.query(models.User).filter_by(id=1).one()
+        mock_request = mock.Mock()
+        mock_request.user = user
+        mock_request.db = self.db
+        mock_request.registry.settings = {
+            'acl_system': acl_system,
+            'pagure_url': 'http://domain.local',
+            'admin_packager_groups': 'provenpackager',
+            'mandatory_packager_groups': 'packager'
+        }
+        mock_request.errors = Errors()
+        mock_request.validated = {'update': update}
+        mock_request.buildinfo = {'bodhi-2.0-1.fc17': {}}
+        return mock_request
+
+    # Mocking the get_pkg_committers_from_pagure function because it will
+    # simplify the overall number of mocks. This function is tested on its own
+    # elsewhere.
+    @mock.patch('bodhi.server.models.Package.get_pkg_committers_from_pagure',
+                return_value=(['guest'], []))
+    def test_validate_acls_pagure(self, mock_gpcfp):
+        """ Test validate_acls when the acl system is Pagure.
+        """
+        mock_request = self.get_mock_request('pagure')
+        validators.validate_acls(mock_request)
+        assert len(mock_request.errors) == 0, mock_request.errors
+        mock_gpcfp.assert_called_once()
+
+    @mock.patch('bodhi.server.models.Package.get_pkg_committers_from_pagure',
+                return_value=(['tbrady'], []))
+    def test_validate_acls_pagure_proven_packager(self, mock_gpcfp):
+        """ Test validate_acls when the acl system is Pagure when the user is
+        a proven packager but doesn't have access through Pagure.
+        """
+        user = self.db.query(models.User).filter_by(id=1).one()
+        group = self.db.query(models.Group).filter_by(
+            name='provenpackager').one()
+        user.groups.pop(0)
+        user.groups.append(group)
+        self.db.flush()
+        mock_request = self.get_mock_request('pagure')
+        validators.validate_acls(mock_request)
+        assert len(mock_request.errors) == 0, mock_request.errors
+        mock_gpcfp.assert_not_called()
+
+    @mock.patch('bodhi.server.models.Package.get_pkg_committers_from_pagure',
+                return_value=(['guest'], []))
+    def test_validate_acls_pagure_not_a_packager(self, mock_gpcfp):
+        """ Test validate_acls when the acl system is Pagure when the user is
+        not a packager but has access through Pagure. This should not be
+        allowed.
+        """
+        user = self.db.query(models.User).filter_by(id=1).one()
+        user.groups.pop(0)
+        self.db.flush()
+        mock_request = self.get_mock_request('pagure')
+        validators.validate_acls(mock_request)
+        error = [{
+            'location': 'body',
+            'name': 'builds',
+            'description': ('guest is not a member of "packager", which is a '
+                            'mandatory packager group')
+        }]
+        assert mock_request.errors == error, mock_request.errors
+        mock_gpcfp.assert_not_called()
+
+    @mock.patch('bodhi.server.models.Package.get_pkg_committers_from_pagure',
+                return_value=(['tbrady'], []))
+    def test_validate_acls_pagure_no_commit_access(self, mock_gpcfp):
+        """ Test validate_acls when the acl system is Pagure when the user is
+        a packager but doesn't have access through Pagure.
+        """
+        mock_request = self.get_mock_request('pagure')
+        validators.validate_acls(mock_request)
+        error = [{
+            'location': 'body',
+            'name': 'builds',
+            'description': 'guest does not have commit access to bodhi'
+        }]
+        assert mock_request.errors == error, mock_request.errors
+        mock_gpcfp.assert_called_once()
+
+    @mock.patch('bodhi.server.models.Package.get_pkg_committers_from_pagure',
+                return_value=(['guest'], []))
+    def test_validate_acls_pagure_runtime_error(self, mock_gpcfp):
+        """ Test validate_acls when the acl system is Pagure and a RuntimeError
+        is raised.
+        """
+        mock_request = self.get_mock_request('pagure')
+        mock_gpcfp.side_effect = RuntimeError('some error')
+        validators.validate_acls(mock_request)
+        assert len(mock_request.errors) == 1, mock_request.errors
+        expected_error = [{
+            'location': 'body',
+            'name': 'builds',
+            'description': 'some error'
+        }]
+        assert mock_request.errors == expected_error, mock_request.errors
+        mock_gpcfp.assert_called_once()
+
+    @mock.patch('bodhi.server.models.Package.get_pkg_committers_from_pagure',
+                return_value=(['guest'], []))
+    def test_validate_acls_pagure_exception(self, mock_gpcfp):
+        """ Test validate_acls when the acl system is Pagure and an exception
+        that isn't a RuntimeError is raised.
+        """
+        mock_request = self.get_mock_request('pagure')
+        mock_gpcfp.side_effect = ValueError('some error')
+        validators.validate_acls(mock_request)
+        assert len(mock_request.errors) == 1, mock_request.errors
+        expected_error = [{
+            'location': 'body',
+            'name': 'builds',
+            'description': ('Unable to access Pagure to check ACLs. Please '
+                            'try again later.')
+        }]
+        assert mock_request.errors == expected_error, mock_request.errors
+        mock_gpcfp.assert_called_once()
+
+    def test_validate_acls_dummy(self):
+        """ Test validate_acls when the acl system is dummy.
+        """
+        mock_request = self.get_mock_request('dummy')
+        validators.validate_acls(mock_request)
+        assert len(mock_request.errors) == 0, mock_request.errors
+
+    def test_validate_acls_invalid_acl_system(self):
+        """ Test validate_acls when the acl system is invalid.
+        This will ensure that the user does not have rights.
+        """
+        mock_request = self.get_mock_request('nonexistent')
+        validators.validate_acls(mock_request)
+        error = [{
+            'location': 'body',
+            'name': 'builds',
+            'description': 'guest does not have commit access to bodhi'
+        }]
+        assert mock_request.errors == error, mock_request.errors

--- a/development.ini.example
+++ b/development.ini.example
@@ -272,8 +272,8 @@ fedmenu.data_url = https://apps.fedoraproject.org/js/data.js
 ##
 ## ACL system
 ## Choices are 'pkgdb', which will send a JSON query to the pkgdb_url below,
-## or 'dummy', which will always return guest credentials (used for local
-## development).
+## 'pagure', which will query the pagure_url below, or 'dummy', which will
+## always return guest credentials (used for local development).
 ##
 acl_system = dummy
 
@@ -281,6 +281,16 @@ acl_system = dummy
 ## Package DB
 ##
 pkgdb_url = https://admin.fedoraproject.org/pkgdb
+
+##
+## Pagure
+##
+# pagure_url = https://src.fedoraproject.org/pagure/
+
+##
+## Product Definition Center (PDC)
+##
+# pdc_url = https://pdc.fedoraproject.org/
 
 # We used to get our package tags from pkgdb, but they come from tagger now.
 # https://github.com/fedora-infra/fedora-tagger/pull/74
@@ -327,11 +337,14 @@ reboot_pkgs = kernel kernel-smp kernel-xen-hypervisor kernel-PAE kernel-xen0 ker
 ## https://fedoraproject.org/wiki/Critical_path_package
 ##
 
-# Enable this to query the Fedora Package Database for the list of Critical
-# Path Packages.  If disabled, it'll just use the hardcoded list below.
+# You can allow Bodhi to query for critpath packages from the Fedora Package
+# Database by setting this value to `pkgdb` or the Product Definition
+# Center by setting this value to `pdc`. If it isn't set, it'll just use the
+# hardcoded list below.
 #critpath.type = pkgdb
 
-# You can hardcode a list of critical path packages instead of using the PackageDB
+# You can hardcode a list of critical path packages instead of using the PkgDB
+# or PDC
 critpath_pkgs = kernel
 
 # The number of admin approvals it takes to be able to push a critical path

--- a/production.ini
+++ b/production.ini
@@ -253,8 +253,8 @@ resultsdb_api_url = https://taskotron.fedoraproject.org/resultsdb_api/
 ##
 ## ACL system
 ## Choices are 'pkgdb', which will send a JSON query to the pkgdb_url below,
-## or 'dummy', which will always return guest credentials (used for local
-## development).
+## 'pagure', which will query the pagure_url below, or 'dummy', which will
+## always return guest credentials (used for local development).
 ##
 acl_system = dummy
 
@@ -262,6 +262,16 @@ acl_system = dummy
 ## Package DB
 ##
 pkgdb_url = https://admin.fedoraproject.org/pkgdb
+
+##
+## Pagure
+##
+# pagure_url = https://src.fedoraproject.org/pagure/
+
+##
+## Product Definition Center (PDC)
+##
+# pdc_url = https://pdc.fedoraproject.org/
 
 # We used to get our package tags from pkgdb, but they come from tagger now.
 # https://github.com/fedora-infra/fedora-tagger/pull/74
@@ -308,11 +318,14 @@ reboot_pkgs = kernel kernel-smp kernel-xen-hypervisor kernel-PAE kernel-xen0 ker
 ## https://fedoraproject.org/wiki/Critical_path_package
 ##
 
-# Enable this to query the Fedora Package Database for the list of Critical
-# Path Packages.  If disabled, it'll just use the hardcoded list below.
+# You can allow Bodhi to query for critpath packages from the Fedora Package
+# Database by setting this value to `pkgdb` or the Product Definition
+# Center by setting this value to `pdc`. If it isn't set, it'll just use the
+# hardcoded list below.
 #critpath.type = pkgdb
 
-# You can hardcode a list of critical path packages instead of using the PackageDB
+# You can hardcode a list of critical path packages instead of using the PkgDB
+# or PDC
 critpath_pkgs = kernel
 
 # The number of admin approvals it takes to be able to push a critical path

--- a/tools/metrics.py
+++ b/tools/metrics.py
@@ -27,7 +27,7 @@ from sqlalchemy.sql import and_
 
 from bodhi.server import Session, initialize_db
 from bodhi.server.models import Update, Release, UpdateStatus, UpdateType
-from bodhi.server.util import header, get_critpath_pkgs
+from bodhi.server.util import header, get_critpath_components
 import bodhi
 
 
@@ -56,7 +56,7 @@ def main(releases=None):
         if releases and release.name not in releases:
             continue
         updates = db.query(Update).filter_by(release=release)
-        critpath_pkgs = get_critpath_pkgs(release.name.lower())
+        critpath_pkgs = get_critpath_components(release.name.lower())
         total = updates.count()
         if not total:
             continue

